Parser Grammar Error Fix: accept function call options only once and disallow unary plus

yjkellyjoo opened the pull request:
This pull request fixes the parser grammar written in ANTLR4 format. Details are as follows:
 (1) function call option should be accepted only once, according to the Solidity v0.8.0 Breaking Changes: New Restrictions (https://docs.soliditylang.org/en/latest/080-breaking-changes.html#new-restrictions)
 (2) unary + operator should be disallowed, according to the Solidity v0.5.0 Breaking Changes: Syntax (https://docs.soliditylang.org/en/latest/050-breaking-changes.html#syntax)

yjkellyjoo committed:
Parser Grammer Error Fix

Fix grammatical error where:
 (1) function call option should be accepted only once, according to the Solidity v0.8.0 Breaking Changes New Restrictions (https://docs.soliditylang.org/en/latest/080-breaking-changes.html#new-restrictions)
 (2) unary + operator should be disallowed, according to the Solidity v0.5.0 Breaking Changes Syntax (https://docs.soliditylang.org/en/latest/050-breaking-changes.html#syntax)

chriseth commented on docs/grammar/SolidityParser.g4:
```diff
@@ -341,11 +341,10 @@ expression:
+	| expression functionCallOption? callArgumentList # FunctionCall
 	| Payable callArgumentList # PayableConversion
 	| Type LParen typeName RParen # MetaType
-	| (Inc | Dec | Not | BitNot | Delete | Sub) expression # UnaryPrefixOperation
```
`Inc` is `++` - what you wanted to disable is `+`, isn't it?

yjkellyjoo replied:
Ah, you're right. I'm currently writing a parser per major compiler versions (0.4.0, 0.5.0, etc.) and did not think to check `Inc` for this version. Will revert this.

chriseth commented on Changelog.md:
```diff
@@ -15,6 +15,7 @@ Bugfixes:
  * Code Generator: Fix internal compiler error when passing a 32-byte hex literal or a zero literal to ``bytes.concat()`` by disallowing such literals.
  * Type Checker: Fix internal error and prevent static calls to unimplemented modifiers.
  * Yul Code Generator: Fix internal compiler error when using a long literal with bitwise negation.
+ * Parser Grammar: Fix grammatical error where (1) function call option should be accepted only once and (2) unary + operator should be disallowed.
```
Maybe "strengthen grammar"?

yjkellyjoo replied:
sounds good, will revise.

yjkellyjoo committed:
Strengthen parser grammar

Fix grammatical error where:
 (1) function call option should be accepted only once, according to the Solidity v0.8.0 Breaking Changes New Restrictions (https://docs.soliditylang.org/en/latest/080-breaking-changes.html#new-restrictions)

chriseth commented:
The tests show that we also need another case for "newExpression". Furthermore, the compiler currently performs the check that call options are only provided once at the type checker level, so maybe we cannot strengthen the grammar in a way that works on all examples in the same way as the compiler does.

yjkellyjoo commented:
I saw that some tests failed on CircleCI but it seems I do not have the permission to browse everything, so couldn't see the detailed code where it failed. Would it be possible to share this case for "newExpression" and maybe also other test codes? I would like to try to make this work.

chriseth commented:
You can run `scripts/test_antlr_grammar.sh` locally to see if it works or not. The new expresion is `new A{value: 2}()`.

yjkellyjoo committed:
Strengthen parser grammar

strengthens the parser grammar where:
 (1) function call option should be accepted only once, according to the Solidity v0.8.0 Breaking Changes New Restrictions (https://docs.soliditylang.org/en/latest/080-breaking-changes.html#new-restrictions)
This new grammar successfully disallows repeated call options demonstrated in test/libsolidity/syntaxTests/functionCalls/calloptions_repeated.sol

yjkellyjoo committed:
Merge branch 'develop' into develop

yjkellyjoo commented:
I have revised the grammar. It should pass the tests now.

chriseth commented:
I'm sorry, but I don't think the reduction in readability through the repetition is worth the gains in better reflecting the actual syntax. There many other details about the syntax that cannot be caught in a context-free grammar.

leonardoalt commented:
@chriseth does that mean you think the PR should be modified? Or just not accepted?
